### PR TITLE
[WIP] Actions and comparison logic

### PR DESCRIPTION
**Pull request by erthalion**
Proof of concept to separate the generation and  application of changes to K8S cluster, and refactoring of comparison logic for stateful sets.

### CODE REVIEW

File: pkg/cluster/sync.go [hunk 1]
```diff
@@ -93,23 +99,42 @@ func (c *Cluster) Sync(newSpec *spec.Postgresql) (err error) {
...
 
-func (c *Cluster) syncServices() error {
+func (c *Cluster) syncServices() (actions []Action, err error) {
 	for _, role := range []PostgresRole{Master, Replica} {
 		c.logger.Debugf("syncing %s service", role)
 
-		if err := c.syncEndpoint(role); err != nil {
-			return fmt.Errorf("could not sync %s endpont: %v", role, err)
+		if err = c.syncEndpoint(role); err != nil {
+			return NoActions, fmt.Errorf("could not sync %s endpont: %v", role, err)
```
Comment (zerg-junior) [inline pkg/cluster/sync.go]: typo in 'endpo*i*nt'

File: pkg/cluster/sync.go [hunk 2]
```diff
@@ -45,8 +46,13 @@ func (c *Cluster) Sync(newSpec *spec.Postgresql) (err error) {
 	}
 
 	c.logger.Debugf("syncing services")
-	if err = c.syncServices(); err != nil {
-		err = fmt.Errorf("could not sync services: %v", err)
+	if actions, err = c.syncServices(); err != nil {
+		err = fmt.Errorf("could not resolve actions to sync services: %v", err)
```
Comment (zerg-junior) [inline pkg/cluster/sync.go]: what does resolve here mean ?

File: pkg/cluster/comparison.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+					return false
+				}
+				return !reflect.DeepEqual(a.Spec.Selector.MatchLabels, b.Spec.Selector.MatchLabels)
+			}, Result{needsReplace: True()}),
+	}
+}
+
+func (c *Cluster) getContainerChecks() []ResourceCheck {
+	return []ResourceCheck{
+		c.NewCheck("new statefulset's container %d name doesn't match the current one",
```
Comment (zerg-junior) [inline pkg/cluster/comparison.go]: will the string interpolation %d happen in all these cases ?

File: pkg/cluster/cluster.go [hunk 4]
```diff
@@ -305,55 +309,13 @@ func (c *Cluster) Create() error {
 
 func (c *Cluster) compareStatefulSetWith(statefulSet *v1beta1.StatefulSet) *compareStatefulsetResult {
 	reasons := make([]string, 0)
-	var match, needsRollUpdate, needsReplace bool
+	update, needsRollUpdate, needsReplace := false, false, false
```
Comment (golangci) [inline pkg/cluster/cluster.go]: ineffectual assignment to `needsRollUpdate`

File: pkg/cluster/cluster.go [hunk 5]
```diff
@@ -362,56 +324,75 @@ func (c *Cluster) compareStatefulSetWith(statefulSet *v1beta1.StatefulSet) *comp
...
-			name := c.Statefulset.Spec.VolumeClaimTemplates[i].Name
-			needsReplace = true
-			reasons = append(reasons, fmt.Sprintf("new statefulset's volumeClaimTemplates specification for volume %q doesn't match the current one", name))
+
+		if check.statefulSetCondition(c.Statefulset, statefulSet) {
+			if util.IsTrue(check.result.needUpdate) {
+				update = true
+			}
+			if util.IsTrue(check.result.needsRollUpdate) {
+				needsRollUpdate = true
```
Comment (golangci) [inline pkg/cluster/cluster.go]: ineffectual assignment to `needsRollUpdate`

File: pkg/cluster/cluster.go [hunk 6]
```diff
@@ -362,56 +324,75 @@ func (c *Cluster) compareStatefulSetWith(statefulSet *v1beta1.StatefulSet) *comp
...
+			}
+			if util.IsTrue(check.result.needsReplace) {
+				needsReplace = true
```
Comment (golangci) [inline pkg/cluster/cluster.go]: ineffectual assignment to `needsReplace`

File: pkg/cluster/cluster.go [hunk 7]
```diff
@@ -362,56 +324,75 @@ func (c *Cluster) compareStatefulSetWith(statefulSet *v1beta1.StatefulSet) *comp
...
+			}
+
+			reasons = append(reasons, check.reason)
 		}
 	}
 
+	needsRollUpdate, reasons = c.compareContainers(c.Statefulset, statefulSet)
```
Comment (golangci) [inline pkg/cluster/cluster.go]: ineffectual assignment to `reasons`

File: pkg/cluster/sync.go [hunk 8]
```diff
@@ -20,6 +20,7 @@ func (c *Cluster) Sync(newSpec *acidv1.Postgresql) error {
 	var err error
 	c.mu.Lock()
 	defer c.mu.Unlock()
+	var actions []Action = NoActions
```
Comment (golangci) [inline pkg/cluster/sync.go]: ineffectual assignment to `actions`

File: pkg/cluster/cluster.go [hunk 9]
```diff
@@ -619,8 +597,10 @@ func (c *Cluster) Delete() {
 			c.logger.Warningf("could not delete %s endpoint: %v", role, err)
 		}
 
-		if err := c.deleteService(role); err != nil {
-			c.logger.Warningf("could not delete %s service: %v", role, err)
+		if actions, err := c.deleteService(role); err != nil {
+			c.logger.Warningf("could prepare actions to delete %s service: %v", role, err)
+		} else {
+			c.applyActions(actions)
```
Comment (golangci) [inline pkg/cluster/cluster.go]: Error return value of `c.applyActions` is not checked

File: pkg/cluster/actions.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+		Delete(action.name, deleteOptions); err != nil {
+		return err
+	}
+
+	meta.cluster.Services[action.role] = nil
+	return nil
+}
+
+func (action ActionService) SetCluster(client *Cluster) {
+	action.meta.cluster = client
```
[the remark this answers is not in the document]
Comment (alexeyklyukin) [reply pkg/cluster/actions.go]: This should be the method of the metadata; therefore, we will avoid implementing it for each implementation of the Action interface.

File: pkg/cluster/cluster.go [hunk 11]
```diff
@@ -82,10 +82,11 @@ type Cluster struct {
 	currentProcess   Process
 	processMu        sync.RWMutex // protects the current operation for reporting, no need to hold the master mutex
 	specMu           sync.RWMutex // protects the spec for reporting, no need to hold the master mutex
+	dryRunMode       bool
```
Comment (alexeyklyukin) [inline pkg/cluster/cluster.go]: Why is this a property of a cluster? Do we expect to run certain clusters in a dry-run mode but not others (wouldn't it be error prone to do so?). If not, I'd put it to the controller and consult this flag when proceeding to execute specific actions (update, delete, sync cluster).

File: pkg/cluster/comparison.go [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+type ResourceCheck struct {
+	containerCondition   ContainerCondition
+	statefulSetCondition StatefulSetCondition
+	volumeClaimCondition VolumeClaimCondition
+	result               Result
+	reason               string
+}
+
+type Result struct {
+	needUpdate      *bool
```
Comment (alexeyklyukin) [inline pkg/cluster/comparison.go]: why do we need pointers here? We could just default to false. This should also eliminate ugly True, False and isTrue

**Comment by alexeyklyukin** [comment]
@erthalion would it make sense to focus on one type of objects (i.e. Services or Statefulsets) to complete this PR and add others afterwards?

**Comment by erthalion** [comment]
Close this in favor of #385 since it was a throw away prototype

**Pull request closed without merging**